src/btree/bt_sync.c

```diff
@@ -114,13 +114,67 @@ __sync_dup_walk(WT_SESSION_IMPL *session, WT_REF *walk, uint32_t flags, WT_REF *
     return (0);
 }
 
+/*
+ * __sync_load_throttle_init --
+ *     Initialize the page load throttle. The (arbitrary) throttle computation is: allow on average
+ *     one load for each internal page in the tree, or at least each one we visit, but allow
+ *     oversubscribing the average by at most four at any time. Reasoning: it should be proportional
+ *     to the tree size (otherwise for large trees the cleanup will never make much progress), and
+ *     the computation should be simple and incremental.
+ */
+static inline void
+__sync_load_throttle_init(uint32_t *load_throttle)
+{
+    *load_throttle = 4;
+}
+
+/*
+ * __sync_load_throttle_next_int --
+ *     Update the page load throttle for visiting a new internal page.
+ */
+static inline void
+__sync_load_throttle_next_int(uint32_t *load_throttle)
+{
+    (*load_throttle)++;
+}
+
+/*
+ * __sync_load_throttle_check --
+ *     Check if the throttle lets us load a page and update the value accordingly.
+ */
+static inline bool
+__sync_load_throttle_check(uint32_t *load_throttle)
+{
+    if (*load_throttle > 0) {
+        (*load_throttle)--;
+        return true;
+    }
+    return false;
+}
+
+/*
+ * __sync_should_read_obsolete_page --
+ *     Check if we should read in an obsolete page with overflow items in order to re-reconcile it
+ *     and discard it all.
+ */
+static int
+__sync_should_read_obsolete_page(WT_SESSION_IMPL *session, uint32_t *load_throttle)
+{
+    /* If the cache is under stress, don't make it worse. */
+    if (__wt_cache_aggressive(session) || __wt_cache_full(session) || __wt_cache_stuck(session) ||
```

Comment:
I would just use `__wt_cache_stuck()` here, that's what compaction uses.

Reply:
(After discussion offline we decided to keep this set of predicates because it's the same as used in __sync_page_skip below.)